chips/lowrisc/src/hmac.rs

```diff
@@ -0,0 +1,248 @@
+                    regs.msg_fifo.set(d);
+
+                    self.data_index.set(data_idx + 4)
+                }
+            });
+        }
+    }
+
+    pub fn handle_interrupt(&self) {
+        let regs = self.registers;
+        let intrs = regs.intr_state.extract();
+
+        regs.intr_enable
+            .modify(INTR_ENABLE::HMAC_DONE::CLEAR + INTR_ENABLE::HMAC_ERR::CLEAR);
+
+        if intrs.is_set(INTR_STATE::HMAC_DONE) {
+            self.digest_dest.map(|digest| {
+                for i in 0..8 {
+                    let d = match i {
+                        0 => regs.digest0.get().to_ne_bytes(),
+                        1 => regs.digest1.get().to_ne_bytes(),
+                        2 => regs.digest2.get().to_ne_bytes(),
+                        3 => regs.digest3.get().to_ne_bytes(),
+                        4 => regs.digest4.get().to_ne_bytes(),
+                        5 => regs.digest5.get().to_ne_bytes(),
+                        6 => regs.digest6.get().to_ne_bytes(),
+                        7 => regs.digest7.get().to_ne_bytes(),
+                        _ => [0, 0, 0, 0],
+                    };
+                    digest[(i * 4) + 0] = d[0];
+                    digest[(i * 4) + 1] = d[1];
+                    digest[(i * 4) + 2] = d[2];
+                    digest[(i * 4) + 3] = d[3];
+                }
+            });
+
+            self.client.map(|client| {
+                client.hash_done(Ok(self.digest_dest.take().unwrap()));
+            });
+        } else if intrs.is_set(INTR_STATE::FIFO_FULL) {
+            // FIFO is full, we can't do anything
+        } else if intrs.is_set(INTR_STATE::HMAC_ERR) {
+            self.client.map(|client| {
+                client.hash_done(Err(regs.err_code.get()));
+            });
+        }
+    }
+}
+
+impl hil::hmac::HMAC<'a> for Hmac<'a> {
+    fn init(&self) {
+        let regs = self.registers;
+
+        regs.cfg
+            .write(CFG::ENDIAN_SWAP::SET + CFG::SHA_EN::SET + CFG::DIGEST_SWAP::SET);
+    }
+
+    fn deinit(&self) {
+        let regs = self.registers;
+
+        regs.cmd.modify(CMD::START::CLEAR);
+        regs.wipe_secret.set(1 as u32);
+    }
+
+    fn set_client(&'a self, client: &'a dyn hmac::Client<'a>) {
+        self.client.set(client);
+    }
+
+    fn set_key(&self, key: &[u8]) -> ReturnCode {
+        let regs = self.registers;
+
+        for i in 0..8 {
+            let mut k = key[i * 4 + 0] as u32;
+            k |= (key[i * 4 + 1] as u32) << 8;
+            k |= (key[i * 4 + 2] as u32) << 16;
+            k |= (key[i * 4 + 3] as u32) << 24;
+
+            match i {
+                0 => regs.key0.set(k),
+                1 => regs.key1.set(k),
+                2 => regs.key2.set(k),
+                3 => regs.key3.set(k),
+                4 => regs.key4.set(k),
+                5 => regs.key5.set(k),
+                6 => regs.key6.set(k),
+                7 => regs.key7.set(k),
+                _ => {}
+            }
+        }
```

Comment:
Not something that needs to be done here but I noticed that this pattern of how to set the keys is common across aes implementations. Might be useful to eventually put this into a common place.

Reply:
If the `key` and `digest` registers were defined as an array, this pattern would naturally go away, isn't it?
```rust
(0x1C => wipe_secret: WriteOnly<u32>),
(0x20 => key: [WriteOnly<u32>; 8]),
(0x40 => digest: [ReadOnly<u32>; 8]),
(0x60 => msg_length_lower: ReadOnly<u32>),
```